pkg/flowaggregator/certificate.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package flowaggregator
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+
+	"math/big"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/apiserver/certificate"
+)
+
+const (
+	CAConfigMapName       = "flow-aggregator-ca"
+	CAConfigMapKey        = "ca.crt"
+	CAConfigMapNamespace  = "flow-aggregator"
+	ClientSecretNamespace = "kube-system"
+	ClientSecretName      = "flow-aggregator-client-tls"
+)
+
+func generateCACertKey() (*x509.Certificate, *rsa.PrivateKey, []byte, error) {
+	cert := &x509.Certificate{
+		SerialNumber: big.NewInt(2020),
+		Subject: pkix.Name{
+			Organization: []string{"VMware, INC."},
+			Country:      []string{"US"},
+			Locality:     []string{"Palo Alto"},
+		},
+		DNSNames:              []string{flowAggregatorDNSName},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(1, 0, 0),
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+	// generate private key for CA
+	caKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	// generate CA certificate
+	caCert, err := x509.CreateCertificate(rand.Reader, cert, cert, &caKey.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	caPEM := new(bytes.Buffer)
+	pem.Encode(caPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: caCert,
+	})
+
+	return cert, caKey, caPEM.Bytes(), err
+}
+
+func generateCertKey(caCert *x509.Certificate, caKey *rsa.PrivateKey) ([]byte, []byte, error) {
+	cert := &x509.Certificate{
+		SerialNumber: big.NewInt(2020),
+		Subject: pkix.Name{
+			Organization: []string{"VMware, INC."},
+			Country:      []string{"US"},
+			Locality:     []string{"Palo Alto"},
+		},
+		DNSNames:    []string{flowAggregatorDNSName},
```

Comment:
We are creating the client certificate at the flow aggregator side, so we will also need to resolve from flowAggregatorDNSName to get the actual flow aggregator collecting address. 
(reason for creating the client cert at flow aggregator is that we need ca key to sign the certificate)

Reply:
I think I wasn't clear originally. I actually had two questions:

1) I don't think the `DNSNames` field should be set for the client certificate - it only makes sense for the server certificate. Is that not accurate?

2) `flowAggregatorDNSName` is hardcoded to `"flow-aggregator.flow-aggregator.svc"`. However, it's possible for the user to deploy the aggregator to another namespace, or with a different service name, or even outside of cluster. I realize now that we cannot use `flowCollectorAddr` (it's not available to the aggregator, it's an option for the agent), so my comment was misleading. However, I think we need a new option in `flow-aggregator.conf` so that the user can provide its own DNSNames. Of course, the default value (if user doesn't provide a value for the option) should be `"flow-aggregator.flow-aggregator.svc"`. *Alternatively*, you can choose to use `ServerName` in your TLS client `Config` (the one you build her: https://github.com/vmware/go-ipfix/blob/f7baa397690f33728a704a5530a24e7c90f7a17a/pkg/exporter/process.go#L342), and set it to `"flow-aggregator.flow-aggregator.svc"` unconditionally. I recommend the first option as it will not interfere with user-provided certificates if / when we add that support in the future. Most users will deploy the flow aggregator using our YAML manifest anyway and will not use to leverage the new config option I suggested.

Reply:
1. You are right. I should not add DNSNames for the client certificate. Fixed it.
2. Yes it makes sense to give users the flexibility to pass dns name. Adopted first option. Thanks for clarification.